src/rmt.rs

```diff
@@ -0,0 +1,415 @@
+    }
+
+    impl CarrierConfig {
+        pub fn new() -> Self {
+            Default::default()
+        }
+
+        pub fn frequency_hz(mut self, f: u32) -> Self {
+            self.frequency_hz = f;
+            self
+        }
+
+        pub fn carrier_level(mut self, state: PinState) -> Self {
+            self.carrier_level = state;
+            self
+        }
+
+        pub fn idle_level(mut self, state: PinState) -> Self {
+            self.idle_level = state;
+            self
+        }
+
+        pub fn duty_percent(mut self, duty: u8) -> Self {
+            self.duty_percent = duty;
+            self
+        }
+    }
+
+    #[derive(Debug, Copy, Clone, Eq, PartialEq)]
+    pub enum Loop {
+        None,
+        Count(u32),
+        Infinite,
+    }
+
+    pub struct WriterConfig {
+        pub clock_divider: u8,
+        pub mem_block_num: u8,
+        pub carrier: Option<CarrierConfig>,
+        // TODO: `loop` is taken. Maybe can change to repeat even though it doesn't match the IDF.
+        pub looping: Loop,
+        /// Enable and set the signal level on the output if idle.
+        pub idle: Option<PinState>,
+        pub aware_dfs: bool,
+    }
+
+    impl Default for WriterConfig {
+        // Defaults from https://github.com/espressif/esp-idf/blob/master/components/driver/include/driver/rmt.h#L101
+        fn default() -> Self {
+            Self {
+                aware_dfs: false,
+                mem_block_num: 1,
+                clock_divider: 80,
+                looping: Loop::None,
+                carrier: None,
+                idle: Some(PinState::Low),
+            }
+        }
+    }
+
+    impl WriterConfig {
+        pub fn new() -> Self {
+            Self::default()
+        }
+
+        /// Channel can work during APB clock scaling.
+        ///
+        /// When set, RMT channel will take REF_TICK or XTAL as source clock. The benefit is, RMT
+        /// channel can continue work even when APB clock is changing.
+        pub fn aware_dfs(mut self, enable: bool) -> Self {
+            self.aware_dfs = enable;
+            self
+        }
+
+        pub fn mem_block_num(mut self, mem_block_num: u8) -> Self {
+            self.mem_block_num = mem_block_num;
+            self
+        }
+
+        pub fn clock_divider(mut self, divider: u8) -> Self {
+            self.clock_divider = divider;
+            self
+        }
+
+        pub fn looping(mut self, looping: Loop) -> Self {
+            self.looping = looping;
+            self
+        }
+
+        pub fn carrier(mut self, carrier: Option<CarrierConfig>) -> Self {
+            self.carrier = carrier;
+            self
+        }
+
+        pub fn idle(mut self, idle: Option<PinState>) -> Self {
+            self.idle = idle;
+            self
+        }
+    }
+}
+
+pub struct Writer {
+    channel: Channel,
+
+    // This must be ManuallyDrop to ensure that it isn't automatically dropped before the driver is
+    // done using the items.
```

Comment:
Is this a real issue? if you call `stop` on the driver in e.g. `release` (the future `drop` replacement), will the driver continue to transmit?

Reply:
In non blocking mode it should stop accessing the data. So I think it's fine to remove the ManuallyDrop.

ManuallyDrop was an artefact I copied from the old, deleted implementation of rmt.